File: client/GhostClient/Sprites/Effects/LineEffect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Ghost.Core.Sharp2D_API;
using Ghost.Sprites.Map;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Sharp2D;
using Sharp2D.Core.Interfaces;

namespace Ghost.Sprites.Effects
{
    public class LineEffect : IEffect
    {
        private static Random random = new Random();
        public void Begin(int duration, int size, float x, float y, double rotation)
        {
            int count = random.Next(400, 600);
            LineSprite[] sprites = new LineSprite[count];

            for (int i = 0; i < count; i++)
            {
                double range = (size/1000.0) - 0.01;
                double angleToAdd = random.NextDouble()* (range - -range) + -range;

                var sprite = new LineSprite(7789, rotation + angleToAdd, duration) { X = x, Y = y, Rotation = (float)(rotation + angleToAdd) };

                sprites[i] = sprite;
            }
            int cursor = 0;

            int lastSpawn = 0;
            int nextSpawn = 0;
            ILogical[] temp = { null };
            int startTime = Environment.TickCount;

            temp[0] = GhostClient.Ghost.CurrentGhostGame.AddLogical(delegate
            {
                if (Environment.TickCount - startTime >= duration)
                {
                    GhostClient.Ghost.CurrentGhostGame.RemoveLogical(temp[0]);
                    sprites = null;
                    return;
                }
                if (Environment.TickCount - lastSpawn > nextSpawn)
                {
                    int toSpawn = random.Next(40, 80);
                    toSpawn = Math.Min(count - cursor, toSpawn);
                    for (int i = cursor; i < cursor + toSpawn; i++)
                    {
                        GhostClient.Ghost.CurrentGhostGame.AddSprite(sprites[i]);
                    }
                    cursor += toSpawn;
                    lastSpawn = Environment.TickCount;
                    nextSpawn = random.Next(10, 100);
                }
            });
        }

        public sealed class LineSprite : Entity
        {
            private int duration;
            private double directoin;
            private int startTime;

            private bool didHit;
            public LineSprite(short id, double rotation, int baseDuration)
                : base(id)
            {
                this.directoin = rotation;
                this.duration = random.Next(baseDuration, (int) (baseDuration*1.5));
                int speed = random.Next(30, 80);
                XVel = (float) (Math.Cos(directoin)*speed);
                YVel = (float) (Math.Sin(directoin)*speed);
                TargetX = 9999f;
                TargetY = 9999f;
                BlendMode = BlendState.Additive;
            }

            protected override void OnLoad()
            {
                Texture = GhostClient.Ghost.CurrentGhostGame.Content.Load<Texture2D>("sprites/ball.png");

                Width = Texture.Width;
                Height = Texture.Height;

                NeverClip = true;
                UniformScale = (float)(random.NextDouble() * (0.35f - 0.2f) + 0.2f);
                TintColor = System.Drawing.Color.FromArgb(255, 194, 19, 19);
            }

            protected override void OnUnload()
            {
            }

            protected override void OnDispose()
            {
            }

            protected override void OnDisplay()
            {
                base.OnDisplay();
                startTime = Environment.TickCount;
            }

            public override void Update()
            {
                base.Update();
                float newAlpha = MathUtils.Ease(1f, 0f, duration, (Environment.TickCount - startTime));

                Alpha = newAlpha;

                if (Alpha == 0f)
                {
                    GhostClient.Ghost.CurrentGhostGame.RemoveSprite(this);
                    return;
                }

                CheckPhysics();
            }

            private void CheckPhysics()
            {
                foreach (IPhysics item in PhysicsObjects.PhysicsItems)
                {
                    Hitbox hitbox = item.Hitbox;

                    if (!Vector2Utils.isPointInside(Position, hitbox.Polygon.Points) && !WillIntersect(hitbox))
                    {
                        continue;
                    }

                    if (hitbox.Name == "WALL")
                    {
                        GhostClient.Ghost.CurrentGhostGame.RemoveSprite(this);
                        return;
                    }

                    Vector2 oldPoint = new Vector2(X - (XVel * 1.5f), Y - (YVel * 1.5f));
                    Vector2 endPoint = new Vector2(X + (XVel * 100), Y + (YVel * 100));

                    Face closestFace = null;
                    Vector2 closestPoint = Vector2.Zero;
                    double distance = 9999999999.0;
                    foreach (Face face in hitbox.Polygon.Faces)
                    {
                        Vector2 pointOfIntersection = Vector2Utils.pointOfIntersection(oldPoint, endPoint, face.PointA,
                            face.PointB);
                        if (pointOfIntersection == Vector2.Zero)
                            continue;

                        double d = Vector2.Distance(pointOfIntersection, oldPoint);
                        if (closestFace == null)
                        {
                            closestFace = face;
                            closestPoint = pointOfIntersection;
                            distance = d;
                        }
                        else if (d < distance)
                        {
                            closestFace = face;
                            closestPoint = pointOfIntersection;
                            distance = d;
                        }
                    }

                    if (closestFace == null)
                        return;

                    if (didHit)
                    {
                        GhostClient.Ghost.CurrentGhostGame.RemoveSprite(this);
                        return;
                    }

                    Vector2 normal = closestFace.Normal;
                    float p = Vector2.Dot(new Vector2(XVel, YVel), normal)*-2f;
                    Vector2 newVel = new Vector2(normal.X, normal.Y);
                    newVel *= p;
                    newVel.X += XVel;
                    newVel.Y += YVel;

                    XVel = newVel.X;
                    YVel = newVel.Y;

                    X = closestPoint.X;
                    Y = closestPoint.Y;

                    didHit = true;
                }
            }

            private bool WillIntersect(Hitbox hitbox)
            {
                Vector2 startPoint = new Vector2(X, Y);
                Vector2 endPoint = new Vector2(X + XVel, Y + YVel);
                int numOfIntersections = 0;

                foreach (Face face in hitbox.Polygon.Faces)
                {
                    Vector2 intersect = Vector2Utils.pointOfIntersection(startPoint, endPoint, face.PointA, face.PointB);
                    if (intersect == Vector2.Zero)
                        continue;
                    numOfIntersections++;
                }

                return numOfIntersections != 0 && numOfIntersections%2 == 0;
            }
        }
    }
}

File: client/core/src/com/boxtrotstudio/ghost/client/core/game/sprites/effects/LineEffect.kt
package com.boxtrotstudio.ghost.client.core.game.sprites.effects

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.TimeUtils
import com.boxtrotstudio.ghost.client.Ghost
import com.boxtrotstudio.ghost.client.core.game.SpriteEntity
import com.boxtrotstudio.ghost.client.core.logic.Logical
import com.boxtrotstudio.ghost.client.core.physics.Face
import com.boxtrotstudio.ghost.client.core.render.Blend
import com.boxtrotstudio.ghost.client.handlers.scenes.SpriteScene
import com.boxtrotstudio.ghost.client.utils.Global
import com.boxtrotstudio.ghost.client.utils.Vector2f
import org.jetbrains.annotations.NotNull

class LineEffect : Effect {
    override fun begin(duration: Int, size: Int, x: Float, y: Float, rotation: Double, world: SpriteScene) {
        val count = Global.rand(400, 600)
        val sprites: Array<LineSprite?> = arrayOfNulls(count)

        for (i in 0..count-1) {
            val range = (size/1000.0)-0.01
            val angleToAdd = Global.RANDOM.nextDouble()*(range - -range)+ -range

            val sprite = LineSprite(rotation + angleToAdd, duration)
            sprite.setCenter(x, y)
            sprite.setRotation(Math.toDegrees(rotation + angleToAdd).toFloat())
            sprite.setBlend(Blend.ADDITIVE);

            sprites[i] = sprite
        }

        var cursor = 0
        var lastSpawn = 0L
        var nextSpawn = 0L
        var startTime = TimeUtils.millis()

        Ghost.getInstance().addLogical(object: Logical {

            override fun tick() {
                if (TimeUtils.millis() - startTime >= duration) {
                    Ghost.getInstance().removeLogical(this)
                    return
                }
                if (TimeUtils.millis() - lastSpawn > nextSpawn) {
                    var toSpawn = Global.rand(40, 80)
                    toSpawn = Math.min(count - cursor, toSpawn)

                    for (i in cursor..cursor + toSpawn-1) {
                        val sprite = sprites[i] ?: continue;
                        world.addEntity(sprite)
                    }

                    cursor += toSpawn
                    lastSpawn = TimeUtils.millis()
                    nextSpawn = (Global.rand(10, 100)).toLong()
                }
            }

            override fun dispose() {

            }
        })
    }
}

class LineSprite(val rotation: Double, val baseDuration: Int) : SpriteEntity("sprites/ball.png", 0) {
    val duration = Global.rand(baseDuration, (baseDuration * 1.5).toInt()).toFloat()
    val speed = Global.rand(30, 70)
    var start = 0f

    override fun onLoad() {
        super.onLoad()

        setZ(1000)
        setScale(Global.RANDOM.nextFloat()*(0.35f - 0.2f)+0.2f)
        color = Color(194 / 255f, 19 / 255f, 19 / 255f, 1f)

        velocity = Vector2f(Math.cos(rotation).toFloat() * speed, Math.sin(rotation).toFloat() * speed)
        target = Vector2f(9999f, 9999f)
    }

    override fun tick() {
        super.tick()

        if (start == 0f) {
            start = TimeUtils.millis().toFloat()
        }

        val newAlpha = ease(1f, 0f, duration, (TimeUtils.millis() - start))

        setAlpha(newAlpha)

        if (newAlpha == 0f) {
            parentScene.removeEntity(this)
            return
        }

        Ghost.PHYSICS.checkEntity(this)
    }

    var didHit = false
    override fun onMirrorHit(closestFace: Face, closestPoint: Vector2f) {
        super.onMirrorHit(closestFace, closestPoint)

        if (!didHit) {
            didHit = true
            velocity.scale(speed.toFloat())
        } else {
            parentScene.removeEntity(this)
        }
    }
}
